Text/Megaparsec/Internal.hs

```diff
@@ -249,12 +254,30 @@ instance (Stream s, MonadIO m) => MonadIO (ParsecT e s m) where
 
 instance (Stream s, MonadReader r m) => MonadReader r (ParsecT e s m) where
   ask = lift ask
-  local f p = mkParsecT $ \s -> local f (runParsecT p s)
+  local f = hoistP (local f)
 
 instance (Stream s, MonadState st m) => MonadState st (ParsecT e s m) where
   get = lift get
   put = lift . put
 
+hoistP ::
+  (Stream s, Monad m) =>
+  (m (Reply e s a) -> m (Reply e s b)) ->
+  ParsecT e s m a ->
```

Comment:
Here, I opted for a less polymorphic version of the signature, since the extra flexibility is not actually needed, so it could confuse future readers (e.g. "do we actually perform a natural tranformation in these cases?").